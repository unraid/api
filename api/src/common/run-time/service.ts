/* eslint-disable new-cap */
import { Number, String, Boolean, Record, Null, type Static } from 'runtypes';

const uptime = Record({
	timestamp: String.Or(Null),
});

export const Service = Record({
	name: String.Or(Null),
	online: Boolean.Or(Null),
	uptime: Number.Or(uptime).Or(Null),
	version: String.Or(Null),
});

// eslint-disable-next-line @typescript-eslint/no-redeclare
export type Service = Static<typeof Service>;
