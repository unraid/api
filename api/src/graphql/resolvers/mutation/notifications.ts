/*!
 * Copyright 2021 Lime Technology Inc. All rights reserved.
 * Written by: Alexis Tyler
 */

import { ensurePermission } from '@app/core/utils/permissions/ensure-permission';
import { graphqlLogger } from '@app/core';
import { GraphQLClient } from '@app/mothership/graphql-client';
import { SEND_NOTIFICATION_MUTATION } from '@app/graphql/mothership/mutations';
import { getters } from '@app/store';
import { type Resolvers } from '@app/graphql/generated/api/types';

export const sendNotification: NonNullable<Resolvers['Mutation']>['sendNotification'] = async (_, args: { notification }, context) => {
	const { user } = context;

	// Check permissions
	ensurePermission(user, {
		resource: 'notifications',
		action: 'create',
		possession: 'own',
	});

	const client = GraphQLClient.getInstance();
	// If there's no mothership connection then bail
	if (!client) {
		graphqlLogger.error('Mothership is not working');
		throw new Error('Mothership is down');
	}
	const result = await client.query({ 
		query: SEND_NOTIFICATION_MUTATION, 
		variables: {
			notification: args.notification,
			apiKey: getters.config().remote.apikey,
		}
	});
	graphqlLogger.debug('Query Result from Notifications.ts', result);
	return args.notification;
};
