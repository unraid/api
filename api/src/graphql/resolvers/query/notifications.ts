import { ensurePermission } from '@app/core/utils/permissions/ensure-permission';
import { type QueryResolvers } from '@app/graphql/generated/api/types';
import { getters } from '@app/store/index';

export const notificationsResolver: QueryResolvers['notifications'] = async (
    _,
    { filter: { offset, limit, importance, type } },
    context
) => {
    ensurePermission(context.user, {
        possession: 'any',
        resource: 'notifications',
        action: 'read',
    });

    if (limit > 50) {
        throw new Error('Limit must be less than 50');
    }
    return Object.values(getters.notifications().notifications)
        .filter((notification) => {
            if (
                importance &&
                importance !== notification.importance
            ) {
                return false;
            }
            if (type && type !== notification.type) {
                return false;
            }

            return true;
        })
        .sort(
            (a, b) =>
                new Date(b.timestamp ?? 0).getTime() -
                new Date(a.timestamp ?? 0).getTime()
        )
        .slice(
            offset,
            limit + offset
        );
};
