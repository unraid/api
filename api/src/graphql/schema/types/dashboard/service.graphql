
type DashboardServiceUptime {
  timestamp: DateTime
}

type DashboardService {
  name: String
  online: Boolean
  uptime: DashboardServiceUptime
  version: String
}

input DashboardServiceUptimeInput {
  timestamp: DateTime!
}

input DashboardServiceInput {
  name: String!
  online: Boolean!
  uptime: DashboardServiceUptimeInput
  version: String!
}