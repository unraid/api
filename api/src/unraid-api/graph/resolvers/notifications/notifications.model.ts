import { Field, InputType, Int, ObjectType, registerEnumType } from '@nestjs/graphql';

import { IsEnum, IsInt, IsNotEmpty, IsOptional, IsString, Min } from 'class-validator';

import { Node } from '@app/unraid-api/graph/resolvers/base.model.js';

export enum NotificationType {
    UNREAD = 'UNREAD',
    ARCHIVE = 'ARCHIVE',
}

export enum NotificationImportance {
    ALERT = 'ALERT',
    INFO = 'INFO',
    WARNING = 'WARNING',
}

// Register enums with GraphQL
registerEnumType(NotificationType, {
    name: 'NotificationType',
});

registerEnumType(NotificationImportance, {
    name: 'NotificationImportance',
});

@InputType('NotificationFilter')
export class NotificationFilter {
    @Field(() => NotificationImportance, { nullable: true })
    @IsEnum(NotificationImportance)
    @IsOptional()
    importance?: NotificationImportance;

    @Field(() => NotificationType)
    @IsEnum(NotificationType)
    @IsNotEmpty()
    type!: NotificationType;

    @Field(() => Int)
    @IsInt()
    @Min(0)
    @IsNotEmpty()
    offset!: number;

    @Field(() => Int)
    @IsInt()
    @Min(1)
    @IsNotEmpty()
    limit!: number;
}

@InputType('NotificationData')
export class NotificationData {
    @Field()
    @IsString()
    @IsNotEmpty()
    title!: string;

    @Field()
    @IsString()
    @IsNotEmpty()
    subject!: string;

    @Field()
    @IsString()
    @IsNotEmpty()
    description!: string;

    @Field(() => NotificationImportance)
    @IsEnum(NotificationImportance)
    @IsNotEmpty()
    importance!: NotificationImportance;

    @Field({ nullable: true })
    @IsString()
    @IsOptional()
    link?: string;
}

@ObjectType('NotificationCounts')
export class NotificationCounts {
    @Field(() => Int)
    @IsInt()
    @Min(0)
    info!: number;

    @Field(() => Int)
    @IsInt()
    @Min(0)
    warning!: number;

    @Field(() => Int)
    @IsInt()
    @Min(0)
    alert!: number;

    @Field(() => Int)
    @IsInt()
    @Min(0)
    total!: number;
}

@ObjectType('NotificationOverview')
export class NotificationOverview {
    @Field(() => NotificationCounts)
    @IsNotEmpty()
    unread!: NotificationCounts;

    @Field(() => NotificationCounts)
    @IsNotEmpty()
    archive!: NotificationCounts;
}

@ObjectType({ implements: () => Node })
export class Notification extends Node {
    @Field({ description: "Also known as 'event'" })
    @Field({ description: "Also known as 'event'" })
    @IsString()
    @IsNotEmpty()
    title!: string;

    @Field()
    @IsString()
    @IsNotEmpty()
    subject!: string;

    @Field()
    @IsString()
    @IsNotEmpty()
    description!: string;

    @Field(() => NotificationImportance)
    @IsEnum(NotificationImportance)
    @IsNotEmpty()
    importance!: NotificationImportance;

    @Field({ nullable: true })
    @IsString()
    @IsOptional()
    link?: string;

    @Field(() => NotificationType)
    @IsEnum(NotificationType)
    @IsNotEmpty()
    type!: NotificationType;

    @Field({ nullable: true, description: 'ISO Timestamp for when the notification occurred' })
    @IsString()
    @IsOptional()
    timestamp?: string;

    @Field({ nullable: true })
    @IsString()
    @IsOptional()
    formattedTimestamp?: string;
}

@ObjectType({ implements: () => Node })
export class Notifications extends Node {
    @Field(() => NotificationOverview, {
        description: 'A cached overview of the notifications in the system & their severity.',
    })
    @IsNotEmpty()
    overview!: NotificationOverview;

    @Field(() => [Notification])
    @IsNotEmpty()
    list!: Notification[];
}
