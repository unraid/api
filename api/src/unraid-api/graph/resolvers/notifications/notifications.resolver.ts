import { type NotificationFilter } from '@app/graphql/generated/api/types';
import { getters } from '@app/store/index';
import { Query, Resolver, Args, Subscription } from '@nestjs/graphql';
import { GraphQLError } from 'graphql';
import { UseRoles } from 'nest-access-control';
import { Logger } from '@nestjs/common';
import { PUBSUB_CHANNEL, createSubscription } from '@app/core/pubsub';

@Resolver()
export class NotificationsResolver {
    @Query()
    @UseRoles({
        resource: 'notifications',
        action: 'read',
        possession: 'any',
    })
    public async notifications(
        @Args('filter')
        { limit, importance, type, offset }: NotificationFilter
    ) {
        if (limit > 50) {
            throw new GraphQLError('Limit must be less than 50');
        }
        return Object.values(getters.notifications().notifications)
            .filter((notification) => {
                if (importance && importance !== notification.importance) {
                    return false;
                }
                if (type && type !== notification.type) {
                    return false;
                }

                return true;
            })
            .sort(
                (a, b) =>
                    new Date(b.timestamp ?? 0).getTime() -
                    new Date(a.timestamp ?? 0).getTime()
            )
            .slice(offset, limit + offset);
    }

    @Subscription('notificationAdded')
    @UseRoles({
        resource: 'notifications',
        action: 'read',
        possession: 'any',
    })
    async notificationAdded() {
        return createSubscription(PUBSUB_CHANNEL.NOTIFICATION);
    }
}
