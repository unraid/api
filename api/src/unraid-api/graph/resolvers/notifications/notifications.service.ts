import { NotificationIni } from '@app/core/types/states/notification';
import { parseConfig } from '@app/core/utils/misc/parse-config';
import { NotificationSchema } from '@app/graphql/generated/api/operations';
import {
    Importance,
    NotificationType,
    type Notification,
    NotificationFilter,
    NotificationOverview,
    NotificationData,
} from '@app/graphql/generated/api/types';
import { getters } from '@app/store';
import { Injectable } from '@nestjs/common';
import { readdir, rename, unlink, writeFile } from 'fs/promises';
import { join } from 'path';
import { Logger } from '@nestjs/common';
import { isFulfilled, isRejected } from '@app/utils';
import { FSWatcher, watch } from 'chokidar';
import { FileLoadStatus } from '@app/store/types';
import { pubsub, PUBSUB_CHANNEL } from '@app/core/pubsub';
import { fileExists } from '@app/core/utils/files/file-exists';
import { encode as encodeIni } from 'ini';

@Injectable()
export class NotificationsService {
    private logger = new Logger(NotificationsService.name);
    private static watcher: FSWatcher | null = null;

    private static overview: NotificationOverview = {
        unread: {
            alert: 0,
            info: 0,
            warning: 0,
            total: 0,
        },
        archive: {
            alert: 0,
            info: 0,
            warning: 0,
            total: 0,
        },
    };

    constructor() {
        NotificationsService.watcher = this.getNotificationsWatcher();
    }

    /**------------------------------------------------------------------------
     *                           Subscription Events
     *
     * Sets up a notification watcher, which hooks up notification lifecycle
     * events to their event handlers.
     *------------------------------------------------------------------------**/

    private getNotificationsWatcher() {
        const { notify, status } = getters.dynamix();
        if (status === FileLoadStatus.LOADED && notify?.path) {
            if (NotificationsService.watcher) {
                return NotificationsService.watcher;
            }

            NotificationsService.watcher = watch(notify.path, {}).on('add', (path) => {
                void this.handleNotificationAdd(path).catch((e) => this.logger.error(e));
            });
            // Do we even want to listen to removals?
            // .on('unlink', (path) => {
            //     void this.handleNotificationRemoval(path).catch((e) =>
            //         this.logger.error(e)
            //     );
            // });

            return NotificationsService.watcher;
        }
        return null;
    }

    private async handleNotificationAdd(path: string) {
        // The path looks like /{notification base path}/{type}/{notification id}
        const type = path.includes('/unread/') ? NotificationType.UNREAD : NotificationType.ARCHIVE;
        this.logger.debug(`Adding ${type} Notification: ${path}`);
        const notification = await this.loadNotificationFile(path, NotificationType[type]);

        NotificationsService.overview[type.toLowerCase()][notification.importance.toLowerCase()] += 1;
        NotificationsService.overview[type.toLowerCase()]['total'] += 1;

        pubsub.publish(PUBSUB_CHANNEL.NOTIFICATION_ADDED, {
            notificationAdded: notification,
        });

        pubsub.publish(PUBSUB_CHANNEL.NOTIFICATION_OVERVIEW, {
            notificationsOverview: NotificationsService.overview,
        });
    }

    private async addToOverview(notification: Notification) {
        const { type, importance } = notification;

        NotificationsService.overview[type.toLowerCase()][importance.toLowerCase()] += 1;
        NotificationsService.overview[type.toLowerCase()]['total'] += 1;

        pubsub.publish(PUBSUB_CHANNEL.NOTIFICATION_ADDED, {
            notificationAdded: notification,
        });

        pubsub.publish(PUBSUB_CHANNEL.NOTIFICATION_OVERVIEW, {
            notificationsOverview: NotificationsService.overview,
        });
    }

    private async removeFromOverview(notification: Notification) {
        const { type, id, importance } = notification;
        this.logger.debug(`Removing ${type} Notification: ${id}`);

        NotificationsService.overview[type.toLowerCase()][importance.toLowerCase()] -= 1;
        NotificationsService.overview[type.toLowerCase()]['total'] -= 1;

        return pubsub.publish(PUBSUB_CHANNEL.NOTIFICATION_OVERVIEW, {
            notificationsOverview: NotificationsService.overview,
        });
    }

    public async getOverview(): Promise<NotificationOverview> {
        return NotificationsService.overview;
    }

    /**------------------------------------------------------------------------
     *                           CRUD: Creating Notifications
     *------------------------------------------------------------------------**/

    public async createNotification(data: NotificationData): Promise<Notification> {
        // const id: string = this.makeNotificationId();
        const id: string = '_DEV_CUSTOM_NOTIFICATION_1234.notify'; // placeholder
        const path = join(this.paths().UNREAD, id);

        const fileData = this.makeNotificationFileData(data);
        const ini = encodeIni(fileData);

        await writeFile(path, ini);
        // await this.addToOverview(notification);
        // make sure both NOTIFICATION_ADDED and NOTIFICATION_OVERVIEW are fired
        return { ...data, id, type: NotificationType.UNREAD, timestamp: fileData.timestamp };
    }

    private makeNotificationFileData(notification: NotificationData): NotificationIni {
        const { title, subject, description, link, importance } = notification;
        const secondsSinceUnixEpoch = Math.floor(Date.now() / 1_000);

        const data: NotificationIni = {
            timestamp: secondsSinceUnixEpoch.toString(),
            event: title,
            subject,
            description,
            importance: this.gqlImportanceToFileImportance(importance),
        };

        // HACK - the ini encoder stringifies all fields defined on the object, even if they're undefined.
        // this results in a field like "link=undefined" in the resulting ini string.
        // So, we only add a link if it's defined

        if (link) {
            data.link = link;
        }
        return data;
    }

    /**------------------------------------------------------------------------
     *                           CRUD: Deleting Notifications
     *------------------------------------------------------------------------**/

    public async deleteNotification({ id, type }: Pick<Notification, 'id' | 'type'>) {
        const path = join(this.paths()[type], id);

        // we don't want to update the overview stats if the deletion (unlink) fails
        // so we do the file system ops first

        const notification = await this.loadNotificationFile(path, type);
        await unlink(path);
        await this.removeFromOverview(notification);

        // return both the overview & the deleted notification
        // this helps us reference the deleted notification in-memory if we want
        return { notification, overview: NotificationsService.overview };
    }

    /**------------------------------------------------------------------------
     *                           CRUD: Updating Notifications
     *------------------------------------------------------------------------**/

    public async archiveNotification({ id }: Pick<Notification, 'id'>): Promise<NotificationOverview> {
        const { UNREAD, ARCHIVE } = this.paths();
        const unreadPath = join(UNREAD, id);
        const archivePath = join(ARCHIVE, id);

        if (!(await fileExists(unreadPath))) {
            this.logger.warn(`[archiveNotification] Could not find notification in unreads: ${id}`);
            return NotificationsService.overview;
        }

        const notification = await this.loadNotificationFile(unreadPath, NotificationType.UNREAD);
        await rename(unreadPath, archivePath);
        await this.removeFromOverview(notification);

        /**-----------------------
         *     Event & PubSub logic
         *
         * We assume `rename` kicks off 'unlink' and 'add' events
         * in the chokidar file watcher.
         *
         * We assume the 'add' handler publishes to
         * NOTIFICATION_ADDED & NOTIFICATION_OVERVIEW,
         * and that no pubsub or overview updates occur upon 'unlink'.
         *
         * Thus, we explicitly update our state & pubsub via `removeFromOverview`
         * and implicitly expect it to be updated via our filesystem changes.
         *
         * The reasons for this discrepancy are:
         *  - Backwards compatibility: not every notification will be created through this API,
         *    so we track state by watching the store (i.e. the file system).
         *
         *  - Technical Limitations: By the time the unlink event fires, the notification file
         *    can no longer be read. This means we can only track overview totals accurately;
         *    to track other stats, we have to update them manually, prior to file deletion.
         *------------------------**/

        return NotificationsService.overview;
    }

    public async markAsUnread({ id }: Pick<Notification, 'id'>) {
        const { UNREAD, ARCHIVE } = this.paths();
        const unreadPath = join(UNREAD, id);
        const archivePath = join(ARCHIVE, id);

        if (!(await fileExists(archivePath))) {
            this.logger.warn(`[markAsUnread] Could not find notification in archive: ${id}`);
            return NotificationsService.overview;
        }

        const notification = await this.loadNotificationFile(archivePath, NotificationType.ARCHIVE);
        await rename(archivePath, unreadPath);

        await this.removeFromOverview(notification);

        // @see `archiveNotification` for why this is commented out, and why there are 2
        //   different ways of updating the overview & pubsub
        //
        // await this.addToOverview({ ...notification, type: NotificationType.UNREAD });

        return NotificationsService.overview;
    }

    /**------------------------------------------------------------------------
     *                           CRUD: Reading Notifications
     *------------------------------------------------------------------------**/

    /**
     * Retrieves all notifications from the file system.
     * @param filters Filters to apply to the notifications
     * @returns An array of all notifications in the system.
     */
    public async getNotifications(filters: NotificationFilter): Promise<Notification[]> {
        this.logger.debug('Getting Notifications');

        const { ARCHIVE, UNREAD } = this.paths();
        const directoryPath = filters.type === NotificationType.ARCHIVE ? ARCHIVE : UNREAD;

        const unreadFiles = await this.getFilesInFolder(directoryPath);
        const [notifications] = await this.getNotificationsFromPaths(unreadFiles, filters);

        return notifications;
    }

    /**
     * Given a path to a folder, returns the full (absolute) paths of the folder's top-level contents.
     * @param folderPath The path of the folder to read.
     * @returns A list of absolute paths of all the files and contents in the folder.
     */
    private async getFilesInFolder(folderPath: string): Promise<string[]> {
        const contents = await readdir(folderPath);

        return contents.map((content) => join(folderPath, content));
    }

    /**
     * Given a an array of files, reads and filters all the files in the directory,
     * and attempts to parse each file as a Notification.
     *
     * Returns an array of two elements:
     * - the first element is an array of successfully parsed and filtered Notifications,
     * - the second element is an array of errors for any files that failed parsing.
     *
     * @param files the files (absolute paths) to read
     * @param filters the filters to apply to the notifications
     * @returns an array of two elements: [successes, errors/failures]
     */
    private async getNotificationsFromPaths(
        files: string[],
        filters: NotificationFilter
    ): Promise<[Notification[], unknown[]]> {
        const { limit, importance, type, offset } = filters;

        const fileReads = files
            .slice(offset, limit + offset)
            .map((file) => this.loadNotificationFile(file, type ?? NotificationType.UNREAD));
        const results = await Promise.allSettled(fileReads);

        return [
            results
                .filter(isFulfilled)
                .map((result) => result.value)
                .filter((notification) => {
                    if (importance && importance !== notification.importance) {
                        return false;
                    }

                    if (type && type !== notification.type) {
                        return false;
                    }

                    return true;
                })
                .sort(
                    (a, b) => new Date(b.timestamp ?? 0).getTime() - new Date(a.timestamp ?? 0).getTime()
                ),
            results.filter(isRejected).map((result) => result.reason),
        ];
    }

    /**
     * Loads a notification file from disk, parses it to a Notification object, and
     * validates the object against the NotificationSchema.
     *
     * @param path The path to the notification file on disk.
     * @param type The type of the notification that is being loaded.
     * @returns A parsed Notification object, or throws an error if the object is invalid.
     * @throws An error if the object is invalid (doesn't conform to the graphql NotificationSchema).
     */
    private async loadNotificationFile(path: string, type: NotificationType): Promise<Notification> {
        const notificationFile = parseConfig<NotificationIni>({
            filePath: path,
            type: 'ini',
        });

        this.logger.debug(
            `Loaded notification ini file from ${path}: ${JSON.stringify(notificationFile, null, 4)}`
        );

        const notification: Notification = {
            id: path,
            title: notificationFile.event,
            subject: notificationFile.subject,
            description: notificationFile.description ?? '',
            importance: this.fileImportanceToGqlImportance(notificationFile.importance),
            link: notificationFile.link,
            timestamp: this.parseNotificationDateToIsoDate(notificationFile.timestamp),
            type,
        };

        // The contents of the file, and therefore the notification, may not always be a valid notification.
        // so we parse it through the schema to make sure it is

        return NotificationSchema().parse(notification);
    }

    private fileImportanceToGqlImportance(importance: NotificationIni['importance']): Importance {
        switch (importance) {
            case 'alert':
                return Importance.ALERT;
            case 'warning':
                return Importance.WARNING;
            default:
                return Importance.INFO;
        }
    }

    private gqlImportanceToFileImportance(importance: Importance): NotificationIni['importance'] {
        switch (importance) {
            case Importance.ALERT:
                return 'alert';
            case Importance.WARNING:
                return 'warning';
            default:
                return 'normal';
        }
    }

    private parseNotificationDateToIsoDate(unixStringSeconds: string | undefined): string | null {
        if (unixStringSeconds && !isNaN(Number(unixStringSeconds))) {
            return new Date(Number(unixStringSeconds) * 1_000).toISOString();
        }
        return null;
    }

    /**
     * Returns the paths to the notification directories.
     *
     * @returns an object with the:
     *          - base path,
     *          - path to the unread notifications,
     *          - path to the archived notifications
     */
    private paths(): Record<'basePath' | NotificationType, string> {
        const basePath = getters.dynamix().notify!.path;
        const makePath = (type: NotificationType) => join(basePath, type.toLowerCase());
        return {
            basePath,
            [NotificationType.UNREAD]: makePath(NotificationType.UNREAD),
            [NotificationType.ARCHIVE]: makePath(NotificationType.ARCHIVE),
        };
    }
}
