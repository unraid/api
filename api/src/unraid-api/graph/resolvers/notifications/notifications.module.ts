import { Module } from '@nestjs/common';

import { NotificationsService } from '@app/unraid-api/graph/resolvers/notifications/notifications.service.js';

@Module({
    providers: [NotificationsService],
    exports: [NotificationsService],
})
export class NotificationsModule {}
