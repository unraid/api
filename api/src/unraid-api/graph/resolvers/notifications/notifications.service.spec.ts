import { Test, TestingModule } from '@nestjs/testing';
import { NotificationsService } from './notifications.service';
import { describe, beforeEach, it, expect, vi, beforeAll, afterAll } from 'vitest';
import { mkdir, rm } from 'fs/promises';
import { existsSync } from 'fs';
import {
    Importance,
    type NotificationData,
    NotificationType,
    type Notification,
} from '@app/graphql/generated/api/types';

describe('NotificationsService', () => {
    let service: NotificationsService;
    const basePath = '/tmp/test/notifications';
    const testPaths = {
        basePath,
        UNREAD: `${basePath}/unread`,
        ARCHIVE: `${basePath}/archive`,
    };

    /**------------------------------------------------------------------------
     *                           Lifecycle Setup
     *------------------------------------------------------------------------**/

    beforeAll(async () => {
        const module: TestingModule = await Test.createTestingModule({
            providers: [NotificationsService],
        }).compile();

        service = module.get<NotificationsService>(NotificationsService); // this might need to be a module.resolve instead of get
        vi.spyOn(service, 'paths').mockImplementation(() => testPaths);

        // clear the notifications directory
        await mkdir(testPaths.basePath, { recursive: true });
        await rm(testPaths.basePath, { force: true, recursive: true });

        // init notification directories
        await mkdir(testPaths.UNREAD, { recursive: true });
        await mkdir(testPaths.ARCHIVE, { recursive: true });
    });

    afterAll(async () => {
        // clear notifications directory
        await rm(testPaths.basePath, { force: true, recursive: true });
    });

    /**------------------------------------------------------------------------
     *                           Helper Functions
     *------------------------------------------------------------------------**/

    async function createNotification(data: Partial<NotificationData> = {}) {
        const {
            title = 'Test Notification',
            subject = 'Test Subject',
            description = 'Test Description',
            importance = Importance.INFO,
        } = data;
        return service.createNotification({ title, subject, description, importance });
    }

    async function findById(id: string, type: NotificationType = NotificationType.UNREAD) {
        return (await service.getNotifications({ type, limit: 50, offset: 0 })).find(
            (notification) => notification.id === id
        );
    }

    async function doesExist(
        { id }: Pick<Notification, 'id'>,
        type: NotificationType = NotificationType.UNREAD
    ) {
        const storedNotification = await findById(id, type);
        expect(storedNotification).toBeDefined();
        return !!storedNotification;
    }

    async function forEachImportance(action: (importance: Importance) => Promise<void>) {
        for (const importance of Object.values(Importance)) {
            await action(importance);
        }
    }

    async function forEachType(action: (type: NotificationType) => Promise<void>) {
        for (const type of Object.values(NotificationType)) {
            await action(type);
        }
    }

    async function forAllTypesAndImportances(
        action: (type: NotificationType, importance: Importance) => Promise<void>
    ) {
        await forEachType(async (type) => {
            await forEachImportance(async (importance) => {
                await action(type, importance);
            });
        });
    }

    /**------------------------------------------------------------------------
     *                           Sanity Tests
     *------------------------------------------------------------------------**/

    it('NotificationsService test setup should be defined', () => {
        expect(service).toBeDefined();
        expect(service.paths()).toEqual(testPaths);
        Object.values(testPaths).forEach((path) => expect(existsSync(path)).toBeTruthy());
    });

    it('can correctly create, load, and delete a notification', async () => {
        const notificationData: NotificationData = {
            title: 'Test Notification',
            subject: 'Test Subject',
            description: 'Test Description',
            importance: Importance.INFO,
        };
        const notification = await createNotification(notificationData);

        // data in returned notification matches?
        Object.entries(notificationData).forEach(([key, value]) => {
            expect(notification[key]).toEqual(value);
        });

        // data in stored notification matches?
        const storedNotification = await findById(notification.id);
        expect(storedNotification).toBeDefined();
        if (!storedNotification) {
            return;
        }

        expect(storedNotification.id).toEqual(notification.id);
        expect(storedNotification.timestamp).toEqual(notification.timestamp);
        Object.entries(notificationData).forEach(([key, value]) => {
            expect(storedNotification[key]).toEqual(value);
        });

        // notification was deleted
        await service.deleteNotification({ id: notification.id, type: NotificationType.UNREAD });
        const deleted = await findById(notification.id);
        expect(deleted).toBeUndefined();
    });

    it('can correctly archive and unarchive a notification', async () => {
        await forEachImportance(async (importance) => {
            const notification = await createNotification({ importance });

            await service.archiveNotification(notification);
            let exists = await doesExist(notification, NotificationType.ARCHIVE);
            if (!exists) return;

            await service.markAsUnread(notification);
            exists = await doesExist(notification, NotificationType.UNREAD);
        });
    });

    it('can archive & unarchive all', async () => {
        await forEachImportance(async (importance) => {
            const notifications = await Promise.all([
                createNotification({ importance }),
                createNotification({ importance }),
                createNotification({ importance }),
                createNotification({ importance }),
                createNotification({ importance }),
            ]);

            expect(notifications.length).toEqual(5);
            await service.archiveAll();
            await service.unarchiveAll();
            await service.archiveAll(importance);
            await service.unarchiveAll(importance);
        });
    });
});
