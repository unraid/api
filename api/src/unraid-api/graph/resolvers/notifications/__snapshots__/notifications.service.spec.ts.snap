// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationsService legacy script compatibility > yields correct cli args for alerts 1`] = `"/usr/local/emhttp/webGui/scripts/notify -i alert -e 'Test Notification !@#$%^&*()_+={}[]|:;"'\\''<>,.?/~\`' -s 'Test Subject \\t\\n🚀💻🛠️' -d 'Test Description with special characters \\t\\n©®™✓✓✓—“”‘’' -l 'https://unraid.net/?query=param&special=💡🔥✨'"`;

exports[`NotificationsService legacy script compatibility > yields correct cli args for alerts 2`] = `
"Command failed with ENOENT: /usr/local/emhttp/webGui/scripts/notify -i alert -e 'Test Notification !@#$%^&*()_+={}[]|:;"'\\''<>,.?/~\`' -s 'Test Subject \\t\\n🚀💻🛠️' -d 'Test Description with special characters \\t\\n©®™✓✓✓—“”‘’' -l 'https://unraid.net/?query=param&special=💡🔥✨'
spawn /usr/local/emhttp/webGui/scripts/notify ENOENT"
`;

exports[`NotificationsService legacy script compatibility > yields correct cli args for normals 1`] = `"/usr/local/emhttp/webGui/scripts/notify -i normal -e 'Test Notification !@#$%^&*()_+={}[]|:;"'\\''<>,.?/~\`' -s 'Test Subject \\t\\n🚀💻🛠️' -d 'Test Description with special characters \\t\\n©®™✓✓✓—“”‘’' -l 'https://unraid.net/?query=param&special=💡🔥✨'"`;

exports[`NotificationsService legacy script compatibility > yields correct cli args for normals 2`] = `
"Command failed with ENOENT: /usr/local/emhttp/webGui/scripts/notify -i normal -e 'Test Notification !@#$%^&*()_+={}[]|:;"'\\''<>,.?/~\`' -s 'Test Subject \\t\\n🚀💻🛠️' -d 'Test Description with special characters \\t\\n©®™✓✓✓—“”‘’' -l 'https://unraid.net/?query=param&special=💡🔥✨'
spawn /usr/local/emhttp/webGui/scripts/notify ENOENT"
`;

exports[`NotificationsService legacy script compatibility > yields correct cli args for warnings 1`] = `"/usr/local/emhttp/webGui/scripts/notify -i warning -e 'Test Notification !@#$%^&*()_+={}[]|:;"'\\''<>,.?/~\`' -s 'Test Subject \\t\\n🚀💻🛠️' -d 'Test Description with special characters \\t\\n©®™✓✓✓—“”‘’' -l 'https://unraid.net/?query=param&special=💡🔥✨'"`;

exports[`NotificationsService legacy script compatibility > yields correct cli args for warnings 2`] = `
"Command failed with ENOENT: /usr/local/emhttp/webGui/scripts/notify -i warning -e 'Test Notification !@#$%^&*()_+={}[]|:;"'\\''<>,.?/~\`' -s 'Test Subject \\t\\n🚀💻🛠️' -d 'Test Description with special characters \\t\\n©®™✓✓✓—“”‘’' -l 'https://unraid.net/?query=param&special=💡🔥✨'
spawn /usr/local/emhttp/webGui/scripts/notify ENOENT"
`;
